core/vdbe/sorter.rs

```diff
@@ -1,67 +1,536 @@
+        for record in records.drain(..) {
+            let payload = record.record.get_payload();
+            buf[buf_pos..buf_pos + payload.len()].copy_from_slice(payload);
+            buf_pos += payload.len();
+        }
+
+        #[allow(clippy::arc_with_non_send_sync)]
+        let buffer_ref = Arc::new(RefCell::new(buffer));
+
+        let buffer_ref_copy = buffer_ref.clone();
+        let chunk_io_state_copy = self.io_state.clone();
+        let write_complete = Box::new(move |bytes_written: i32| {
+            chunk_io_state_copy.set(SortedChunkIOState::WriteComplete);
+            let buf_len = buffer_ref_copy.borrow().len();
+            if bytes_written < buf_len as i32 {
+                tracing::error!("wrote({bytes_written}) less than expected({buf_len})");
+            }
+        });
+
+        let c = Completion::new(CompletionType::Write(WriteCompletion::new(write_complete)));
+        self.file.pwrite(0, buffer_ref, c)?;
+        Ok(())
     }
-    pub fn record(&self) -> Option<&ImmutableRecord> {
-        self.current.as_ref()
+}
+
+struct SortableImmutableRecord {
+    record: ImmutableRecord,
+    key_len: usize,
+    order: IndexKeySortOrder,
+    collations: Rc<Vec<CollationSeq>>,
+}
+
+impl SortableImmutableRecord {
+    fn new(
+        record: ImmutableRecord,
+        key_len: usize,
+        order: IndexKeySortOrder,
+        collations: Rc<Vec<CollationSeq>>,
+    ) -> Self {
+        Self {
+            record,
+            key_len,
+            order,
+            collations,
+        }
+    }
+}
+
+impl Ord for SortableImmutableRecord {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let this_values = self.record.get_values();
+        let other_values = other.record.get_values();
+
+        let a_key = if this_values.len() >= self.key_len {
+            &this_values[..self.key_len]
+        } else {
+            &this_values[..]
+        };
+
+        let b_key = if other_values.len() >= self.key_len {
+            &other_values[..self.key_len]
+        } else {
+            &other_values[..]
+        };
+
+        compare_immutable(a_key, b_key, self.order, &self.collations)
+    }
+}
+
+impl PartialOrd for SortableImmutableRecord {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
     }
+}
+
+impl PartialEq for SortableImmutableRecord {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl Eq for SortableImmutableRecord {}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum SortedChunkIOState {
+    WaitingForRead,
+    ReadComplete,
+    ReadEOF,
+    WaitingForWrite,
+    WriteComplete,
+    None,
+}
 
-    pub fn insert(&mut self, record: &ImmutableRecord) {
-        self.records.push(record.clone());
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::types::{ImmutableRecord, RefValue, Value};
+    use crate::PlatformIO;
+
+    #[test]
+    fn test_external_sort() {
```

Comment:
Can we make this a fuzz test using:

1. Random insertion count
2. Random record structure (not just single ints)

Reply:
Should I move this test into the `simulator` trait to benefit from `Arbitrary` implementation for `Table` and `SimValue`? Or do you suggest doing something else?

Reply:
I've made this a fuzz test. Please let me know what you think